### PR TITLE
New environment terminal language

### PR DESCRIPTION
**Pull request by jeromefv**
## PR Summary

Modified terminal command language to indicate that user can control directory for creating new environment.

## PR Checklist

- [ x] Has pytest style unit tests (and `pytest` passes).
- [ x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ x] New features are documented, with examples if plot related.
- [ x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [x ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: doc/README.txt
Comment (jklymak) [inline doc/README.txt:40]:
```suggestion
```

File: doc/devel/contributing.rst
Comment (jklymak) [inline doc/devel/contributing.rst:92]: I think `/path/to/devel/env` is pretty clearly a place holder?

File: doc/devel/contributing.rst
Comment (jklymak) [inline doc/devel/contributing.rst:96]: this needs the `bin/` or it won't work I think

**Review by jklymak — DISMISSED** [review]
Thanks a lot for contributing!  OTOH I'm not convinced that this is an improvement - the placeholder is already pretty clearly just a placeholder that the user should modify.

File: doc/devel/contributing.rst
Comment (story645) [reply doc/devel/contributing.rst:92]: I don't think so since it's windows specific & I've seen it trip up enough users that I switched to <placeholder> in all my tutorials and the like.

Comment (jklymak) [reply doc/devel/contributing.rst:92]: What is linux-specific?  The forward slashes?

Comment (story645) [reply doc/devel/contributing.rst:92]: to  be clearer, I've seen folks not recognize it as a placeholder and literally type in /path/to/dev/env

Comment (jklymak) [reply doc/devel/contributing.rst:92]: ok?  And then it doesn't work because `/path` doesn't exist (any more than `<file folder location>` does) and they realize their mistake, right?

Comment (jeromefv) [reply doc/devel/contributing.rst:92]: As a new user contributing, I have little to no experience in setting up a new environment and took this as something I should execute verbatim. This line of text did not execute as MacOS with Catalina doesn't allow write access to that folder. This change supports a user's ability to make use of the terminal in a way that allows them to have control over where things are and take agency in that executed action.

File: doc/devel/contributing.rst
Comment (jeromefv) [reply doc/devel/contributing.rst:96]: Ah, yes. That addition is necessary for the suggested change. Thank you!

File: doc/devel/contributing.rst
Comment (jklymak) [reply doc/devel/contributing.rst:92]: This is copied almost verbatim from the linked docs.  https://docs.python.org/3/library/venv.html  I think its better to stay consistent with the linked docs.

Comment (story645) [reply doc/devel/contributing.rst:92]: >they realize their mistake, right?
Kinda, usually they go down a google rabbit hole of getting more confused. The path/to/folder formulation introduces more friction... especially on windows where it's path\to\folder.

And venv having confusing docs is not a reason for us to do it.

Comment (anntzer) [reply doc/devel/contributing.rst:92]: you realize forward slashes also work on Windows, right?

Comment (story645) [reply doc/devel/contributing.rst:92]: @anntzer weirdly depending on which terminal you're using...like path completion only works properly if you use slashes in one direction and not the other...and the standard documemted way is slashes in the other direction and honestly like "path/to/dev/" being similar enough to something that could work is the problem in that it makes it harder to recognize that it's a placeholder.

Comment (anntzer) [reply doc/devel/contributing.rst:92]: I'm pretty sure path completion works on both kinds of slashes in both cmd.exe and powershell.

Comment (story645) [reply doc/devel/contributing.rst:92]: > I'm pretty sure path completion works on both kinds of slashes in both cmd.exe and powershell.
I just tested it...
ezgif-7-a965b63e1dd2

File: doc/devel/contributing.rst
Comment (story645) [inline doc/devel/contributing.rst:97]: The windows specific ones should probably have the / go the other way \ as that's what the google style guide recommends and on some dev call we decided that would be a good base reference. https://developers.google.com/style/slashes

**Review by story645 — CHANGES_REQUESTED** [review]
Accidental comment (I blame the app). :(

File: doc/devel/contributing.rst
Comment (anntzer) [reply doc/devel/contributing.rst:92]: Turns out my memory failed me :/

**Comment by jklymak** [comment]
I'd say if someone is blindly typing:

```
$ python3 -mvenv <file folder location>
zsh: parse error near `\n'
$
$ python3 -mvenv /path/to/dev/env
Error: [Errno 30] Read-only file system: '/path'
```

the second error is at least pretty clear what the issue is.  I'll note again that these are *developer* docs.  I think we are allowed to assume some level of expertise in folks who want to set up a development environment.

**Comment by story645** [comment]
on windows, the latter doesn't even flag as an error:
```bat

C:\Users\story\Projects\proposal>python3 -mvenv <file folder location>
The syntax of the command is incorrect.

C:\Users\story\Projects\proposal>python3 -mvenv /path/to/dev/env
```
>  I think we are allowed to assume some level of expertise in folks who want to set up a development environment.

Documentation contributions require setting up a development environment and in particular for some of the docs we want low barriers to entry for contributing. Maybe the solution will end up being that they use github codeblocks but I don't think that's camera ready yet.

**Comment by jklymak** [comment]
> Documentation contributions require setting up a development environment and in particular for some of the docs we want low barriers to entry for contributing.

Why do you need an editable install to edit the documentation?

**Comment by story645** [comment]
> Why do you need an editable install to edit the documentation?

because the documentation is in sphinx/rst with enough matplotlib custom stuff that github's preview doesn't handle it, so once you're doing anything on the scope of narrative docs you're gonna want to be able to build them.

**Comment by jklymak** [comment]
Hah, well if you want to build the docs locally, you've got a long haul after you manage to set up your dev environment.  I *never* build locally, I just let CI do it for me.

**Comment by jeromefv** [comment]
> I'd say if someone is blindly typing:
> 
> ```
> $ python3 -mvenv <file folder location>
> zsh: parse error near `\n'
> $
> $ python3 -mvenv /path/to/dev/env
> Error: [Errno 30] Read-only file system: '/path'
> ```
> 
> the second error is at least pretty clear what the issue is. I'll note again that these are _developer_ docs. I think we are allowed to assume some level of expertise in folks who want to set up a development environment.

This was an error I ran into. It's not obvious to me that the file folder location is something that can change without any other context. As someone who is not a developer, I assumed that I had to gain write access to that specific file folder location. Upon researching that option, it was a very involved solution for a task. I only later independently discovered that I could change the file folder location and move forward in the process of building documentation. Searching for solutions for why I wasn't able to get write access only took me further away from the goal of creating a virtual environment in Python. 

I eventually solved my own problem; however, it would have been remedied by a simple language addition and an explanation for it as well. I don't see any conversely catastrophic harm done to the process when incorporating friendlier language aside from cohesion with Python developer docs and possible conflict using the direct copy/paste of terminal commands.

**Comment by jklymak** [comment]
Again, I think this is not an improvement.  But other devs can dismiss my review.

File: doc/devel/contributing.rst
Comment (timhoffm) [reply doc/devel/contributing.rst:92]: I agree with @jeromefv and @story645 that we should not unnecessarily assume user knowledge here. IMHO a battle of `/path/to/devel/env` vs. `<file folder location>` is somewhat futile. It's an improvement, but marginal compared to the overall issue. The whole section (probably page) assumes too much and lacks motivation, context and explanation.

IMHO it should be something like

~~~
Installing Matplotlib for development
-------------------------------------

To decouple Matplotlib development from your working python environment, we
strongly recommend using a dedicated `virtual environment`_ (or alternatively
conda environment).

The virtual environment information is stored in a folder of your choice
(called `<dev env folder>` in the following`). A new environment can be set up
with ::

   python3 -mvenv <dev env folder>

~~~

**Review by timhoffm — COMMENTED** [review]
On a more general note, we need to address both, the experienced user who just needs a brief summary, but also the novice who needs a lot more explanation.

#### Structuring is a key for serving both.
We should reduce long prose sections. One could e.g. summarize the steps at the top of the section with an item list (as an overview and for the experienced user) and/or add subsections explaining the steps in detail. Or use collapsible blocks for more context information (has anybody written a sphinx plugin for this?).

**Comment by brunobeltran** [comment]
> Hah, well if you want to build the docs locally, you've got a long haul after you manage to set up your dev environment. I _never_ build locally, I just let CI do it for me.

Jody do you have any particular issues getting the docs to build? I think that with perfect documentation, an experienced Python developer would be able to set up our docs to build locally in, say 5min (less time than it takes for Sphinx to actually run :/). 

I know that right now we're missing documentation of some particular dependencies, but besides those system-specific brew/apt-get/whatever commands, ideally a `git clone`, followed by an appropriate `pip install -r` and a `make html` should suffice. 


> I think we are allowed to assume some level of expertise in folks who want to set up a development environment.

Obviously there is quite a bit of disagreement among the core team as to what level of expertise should be expected at different places in the docs. As someone who teaches people at many different levels of Python expertise, I would argue that this PR is an improvement as-is, but would recommend against any larger efforts to solve the problem of making docs contributions easier by making setting up a dev environment easier.

> because the documentation is in sphinx/rst with enough matplotlib custom stuff that github's preview doesn't handle it, so once you're doing anything on the scope of narrative docs you're gonna want to be able to build them.

Sphinx is fundamentally tricky. RST is fundamentally tricky. I don't think that we should bend over backwards for the microscopic chance that we entice a novice developer to spontaneously start making large enough changes that they *really* need to see how they render before sending in a PR. And for simpler stuff, instead of funneling anyone that wants to make a doc contribution through setting up the whole dev environment, we should just have them send it into us. 

I for one don't think modifying a well-meaning docs PR with bad RST syntax to be compliant is really much harder than reviewing it in the first place. 

To me, the long term solution can be found with our migration to a new Sphinx theme. As we start to set it up, having (for example) a floating element that says "Found something confusing? Click here to fix it!" that sends them to a page with the instructions for editing that *specific* page ((1) fork, (2) clone, (3) which file to edit, pref with line numbers, (4) simple add/commit/push, (5) click PR button).

**Comment by story645** [comment]
>  And for simpler stuff, instead of funneling anyone that wants to make a doc contribution through setting up the whole dev environment, we should just have them send it into us.

I think the real real solution here is suggest they use GitHub codespaces given the environment builds there. Especially since we have an I think working docker build script for the docs that can probably be configured into a GitHub action.

I agree w/ @timhoffm though that the real issue here is the question of audience & that expectation setting at the top of the doc would be useful.

**Comment by jklymak** [comment]
OT, but, I don't build the docs locally because the dependencies are non trivial, not all listed, and a messy guessing game of `brew` and `conda` on my machine.  Definitely I've done it, but I really don't see the point - it takes forever to build anyways, so why not just let another machine somewhere do it?  Well, other than that is probably bad for the environment.

**Comment by dopplershift** [comment]
"They're dev docs, they're allowed to be complicated" as a general sentiment is a fantastic method to ensure we don't on-board any novice developers. I'm not saying we have to bend over backwards. In this particular case, where we have a demonstrated case that the current content was confusing and we try to clarify in a way that requires minimum additional prose, I'm 👍 .

> I think the real real solution here is suggest they use GitHub codespaces given the environment builds there.

@story645 I don't think you want to rely on codespaces, as here's what they say about pricing:
> Pricing for Codespaces has not yet been finalized. Code-editing functionality in GitHub will always be free. We plan to offer simple pay-as-you-go pricing for Codespaces cloud environments. Codespaces is free during the limited beta.

**Comment by story645** [comment]
@jeromefv the only thing that's holding up approval from me is the slash order for the windows path.

File: doc/devel/contributing.rst
Comment (jeromefv) [reply doc/devel/contributing.rst:97]: > @jeromefv the only thing that's holding up approval from me is the slash order for the windows path.

@story645 I'm not sure if I've got that change showing up anywhere else besides my local branch at the moment. When I've executed `git pull upstream master` and see what changes there are, it appears that those edits have already been put in place. I feel like I'm missing something and I'm not quite sure how to move forward from that.

**Comment by QuLogic** [comment]
Not too sure you did that rebase right.

**Comment by jeromefv** [comment]
@jklymak @story645 @brunobeltran I believe with Bruno's help that I've resolved these issues! Thank you to everyone for helping me get through my first PR successfully. I appreciate all you've done!

**Review by story645 — APPROVED** [review]

File: doc/devel/contributing.rst
Comment (jklymak) [reply doc/devel/contributing.rst:92]: You didn't actually follow @timhoffm 's suggestion to explain what `<file folder location>` is supposed to mean.

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Doc build failure is unrelated to this PR (c.f. #18492).

**Comment by timhoffm** [comment]
@jeromefv Congratulations on your first contribution to Matplotlib! 🎉

**Pull request merged**
